tectonic: move to by-name

bryango opened the pull request:
**Background:**
- there _might_ be a new tectonic release coming up, so I think it would be a good time to do some cleanups beforehand
- after the treewide re-format, the move is mostly trivial

**Details:**
1. the first commit is minimal: it simply moves the package files to `by-name` and keeps the entrypoints in `all-packages`
2. the second commit is a preparation for the third one: it makes `texpresso` depends on `tectonic.unwrapped` (dot) instead of `tectonic-unwrapped`(dash); this is necessary for the next commit
3. the third commit moves `tectonic-unwrapped` and `biber-for-tectonic` into the more manageable `aliases.nix`:


* `biber-for-tectonic -> tectonic.biber`
* `tectonic-unwrapped -> tectonic.unwrapped`

... therefore dropping all refs to tectonic in `all-packages`

**Note:** 
- **I believe the 2nd and 3rd commits (i.e. dropping from `all-packages`) make it more consistent with `pkgs/by-name`, but if it feels too involved, I am fine with keeping only the 1st commit (i.e. keeping the entrypoints in `all-packages`). Moreover:**

- The 2nd and the 3rd commits also have a downside: it seems that moving to aliases will make the dependency packages (`tectonic-unwrapped` and `biber-for-tectonic`) invisible to consumers; [they no longer show up in hydra evals](https://github.com/NixOS/nixpkgs/actions/runs/13302608858/attempts/1#summary-37146864727), so probably also not on search.nixos.org.

- Another option is to create:
	* `pkgs/by-name/te/tectonic-unwrapped` and
	* `pkgs/by-name/bi/biber-for-tectonic`

	... but then the nix source would be split and scattered into different directories. If this is more desirable, I can also implement it.

- Darwin ofborg failures seem unrelated; see https://github.com/NixOS/ofborg/issues/704. Ofborg happy on [x86_64-linux](https://github.com/NixOS/nixpkgs/pull/381668/checks?check_run_id=37148125045).

- This patch should lead to _zero_ rebuild; tested locally by checking these derivation paths:

```console
$ for attr in tectonic{,-unwrapped} tectonic.{unwrapped,biber} biber-for-tectonic texpresso; do nix eval --file . "$attr"; done
«derivation /nix/store/vaf0wmmvdwv9pz7g5ylwgz5wvkk3q0hm-tectonic-wrapped-0.15.0.drv»
«derivation /nix/store/bbid7fm9ydxlkh0c6mcab7d516s9rw0w-tectonic-0.15.0.drv»
«derivation /nix/store/bbid7fm9ydxlkh0c6mcab7d516s9rw0w-tectonic-0.15.0.drv»
«derivation /nix/store/sa6zaxap6n8gzah7r989y67ydbbdlbf6-perl5.40.0-biber-2.17.drv»
«derivation /nix/store/sa6zaxap6n8gzah7r989y67ydbbdlbf6-perl5.40.0-biber-2.17.drv»
«derivation /nix/store/ykl4gahw143p9qmx971wvgy0f1n8chag-texpresso-0-unstable-2025-01-29.drv»
```

Your hashes may be different because of different `HEAD`s, but the hashes should be the same before and after the patch.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bryango committed:
tectonic: move to by-name

This commit is intentionally minimal, such that git may track the file
renames correctly.

doronbehar reviewed (dismissed):
LGTM! Thanks!

doronbehar commented:
> * The 2nd and the 3rd commits also have a downside: it seems that moving to aliases will make the dependency packages (`tectonic-unwrapped` and `biber-for-tectonic`) invisible to consumers; [they no longer show up in hydra evals](https://github.com/NixOS/nixpkgs/actions/runs/13302608858/attempts/1#summary-37146864727), so probably also not on search.nixos.org.

But the aliases should still have these attributes, so they will become invisible only if someone explicitly disables aliases no?

bryango commented:
> > * The 2nd and the 3rd commits also have a downside: it seems that moving to aliases will make the dependency packages (`tectonic-unwrapped` and `biber-for-tectonic`) invisible to consumers; [they no longer show up in hydra evals](https://github.com/NixOS/nixpkgs/actions/runs/13302608858/attempts/1#summary-37146864727), so probably also not on search.nixos.org.
> 
> But the aliases should still have these attributes, so they will become invisible only if someone explicitly disables aliases no?

Yes, aliases are allowed by default so most people using the `tectonic-unwrapped` and `biber-for-tectonic` will be able to keep using it; we do have backward compatibility! What I mean is that these attributes will stop being _advertised_ on search.nixos.org or hydra, as these services do not evaluate aliases.

doronbehar commented:
Hmm I have to say, on a 2nd thought that I think the current state of `nix search tectonic tex` is very good:

```
* legacyPackages.x86_64-linux.biber-for-tectonic (2.17)
  Backend for BibLaTeX

* legacyPackages.x86_64-linux.tectonic (0.15.0)
  Tectonic TeX/LaTeX engine, wrapped with a compatible biber

* legacyPackages.x86_64-linux.tectonic-unwrapped (0.15.0)
  Modernized, complete, self-contained TeX/LaTeX engine, powered by XeTeX and TeXLive
```

And if the 2 commits at question will change that (not sure to what at the moment), then I think I'd object them. I'm not sure because I just checked with `nix search . tectonic tex` on a checkout of this PR and got the same result.

Either way, I think what's good in the current `nix search` results is that the non-trivial `biber` - `tectonic` match and how it makes it clear that matching those is not trivial, without delving too much into details. Also, it is very clear for the user what kind of options they have.

bryango committed:
tectonic-unwrapped, biber-for-tectonic: link to tectonic.{unwrapped,biber}

We do not use aliases because we would like these attributes to be
discoverable by consumers on search.nixos.org, which currently sets
`allowAliases = false`. This behavior is defined in:

- pkgs/top-level/packages-config.nix

Also we do not want to split and scatter these highly related packages
into far away subdirectories, therefore we simply redirect them to the
expressions defined under `pkgs/by-name/te/tectonic/`.

bryango commented:
> I just checked with `nix search . tectonic tex` on a checkout of this PR and got the same result.

`nix search` would not change because it is evaluated locally with the default `allowAliases = true`, so it would show the exact same result as before (which is good). However, it seems that the packages database on https://search.nixos.org is generated with `allowAliases = false`:

https://github.com/NixOS/nixpkgs/blob/5c6ca3c5aaba4ba12b5e18021745bc65c5999b03/pkgs/top-level/packages-config.nix#L1-L5

So the `tectonic-unwrapped` and `biber-for-tectonic` attributes would not be shown. This is indeed not very nice.

------
I just pushed some new commits experimenting for a better solution: instead of `aliases.nix`, we simply create some stub packages:
- `pkgs/by-name/te/tectonic-unwrapped` and
- `pkgs/by-name/bi/biber-for-tectonic`

and redirect them to `tectonic.{unwrapped,biber}` in their respective `package.nix`. This should resolve the issue.

doronbehar commented on pkgs/by-name/te/tectonic/package.nix:
```diff
@@ -53,6 +53,14 @@ symlinkJoin {
 
   meta = tectonic.unwrapped.meta // {
     description = "Tectonic TeX/LaTeX engine, wrapped with a compatible biber";
+    longDescription =
+      (tectonic.unwrapped.longDescription or "")
+      + ''
+        Note that:
+        - `tectonic` by default provides `tectonic` bundled with the compatible `biber`
+        - `tectonic.biber` is the compatible `biber` executable
+        - `tectonic.unwrapped` provides the unwrapped `tectonic` package
+      '';
```
Hmm OK I guess this is good, but it is slightly confusing that in a search you'd encounter both the mention of `tectonic.{biber,unwrapped}` here and (with or without aliases enabled) also `biber-for-tectonic` and `tectonic-unwrapped`? You'd encounter them both in `nix search` and search.nixos.org . Since `nix search` doesn't show `longDescription`s but does show aliases (by default), and search.nixos.org doesn't show aliases but does show `longDescription`s, perhaps we should go after all with an alias, even with a warning?

bryango replied:
Indeed there is some inconsistency here... Maybe we simply drop this `longDescription`? This now just adds noise to the consumer. With the `pkgs/by-name/` stubs we can achieve full backward compatibility, so we just keep the long names (`tectonic-unwrapped`, `biber-for-tectonic`) as "public interfaces"?

doronbehar replied:
> Indeed there is some inconsistency here... Maybe we simply drop this `longDescription`? This now just adds noise to the consumer. With the `pkgs/by-name/` stubs we can achieve full backward compatibility, so we just keep the long names (`tectonic-unwrapped`, `biber-for-tectonic`) as "public interfaces"?

OK that's fine with me :).

bryango commented:
ghborg happy. Now it reports [0 package removed](https://github.com/NixOS/nixpkgs/actions/runs/13352569989/attempts/1#summary-37291037377) (c.f. previous implementation leads to [2 packages removed](https://github.com/NixOS/nixpkgs/actions/runs/13302608858/attempts/1#summary-37146864727); this is what we are trying to avoid).

bryango commented on pkgs/by-name/bi/biber-for-tectonic/package.nix:
```diff
@@ -0,0 +1,4 @@
+/**
+  refer to `pkgs.tectonic` for the definition
+*/
+{ tectonic }: tectonic.biber
```
By `pkgs/by-name` stubs I mean e.g. this

doronbehar replied:
> By `pkgs/by-name` stubs I mean e.g. this

Yea I understand :) :+1: .

doronbehar commented:
> this is what we are trying to avoid

I didn't necessarily wanted to avoid that particular detail, I'm only concerned with the consistency and clarity of what people will read in the search results.

doronbehar commented:
BTW regarding the CI errors, another option is to put in `all-packages.nix` the lines:

```nix
tectonic-unwrapped = tectonic.unwrapped;
biber-for-tectonic = tectonic.biber;
```

bryango commented:
> BTW regarding the CI errors, another option is to put in `all-packages.nix` the lines:
> 
> ```nix
> tectonic-unwrapped = tectonic.unwrapped;
> biber-for-tectonic = tectonic.biber;
> ```

I have pushed this implementation (indeed, the `pkgs/by-name` stubs are problematic as it breaks `.override`; that would be very confusing for users).

Another option is that we actually split it into three packages; this is probably the cleanest way to do it, but then we need to maintain 3 packages living in 3 different subdirectories... This alternative is implemented in https://github.com/NixOS/nixpkgs/compare/7718dcbc936df8813c17cc0f3986eb3c6d5a0624

I am fine with either solution. Which do you prefer? :laughing:

bryango commented:
The commit message was incorrect; just force-pushed the fix.

doronbehar commented:
It was kind of hard for me to explain in words what I think would be an ideal set of commits, so I created an alternative PR :):

- https://github.com/NixOS/nixpkgs/pull/382520

bryango commented:
Looking good! Closing in favor of https://github.com/NixOS/nixpkgs/pull/382520
